sdk/core/azure-core-http-jdk-httpclient/src/main/java/com/azure/core/http/jdk/httpclient/implementation/InputStreamTimeoutResponseSubscriber.java

```diff
@@ -0,0 +1,309 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.core.http.jdk.httpclient.implementation;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.http.HttpResponse;
+import java.net.http.HttpTimeoutException;
+import java.nio.ByteBuffer;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Objects;
+import java.util.TimerTask;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.Flow;
+import java.util.concurrent.Semaphore;
+
+/**
+ * Implementation of {@link HttpResponse.BodySubscriber} that emits the response body as a {@link InputStream} while
+ * tracking a timeout for each value emitted by the subscription.
+ * <p>
+ * This differs from {@link HttpResponse.BodySubscribers#ofPublisher()} in that it tracks a timeout for each value
+ * emitted by the subscription. This is needed to offer better reliability when reading the response in cases where the
+ * server is either slow to respond or the connection has dropped without a signal from the server.
+ */
+public final class InputStreamTimeoutResponseSubscriber extends InputStream
+    implements HttpResponse.BodySubscriber<InputStream> {
+    // Sentinel values to indicate completion.
+    private static final ByteBuffer LAST_BUFFER = ByteBuffer.wrap(new byte[0]);
+    private static final List<ByteBuffer> LAST_LIST = List.of(LAST_BUFFER);
+
+    // A queue of yet unprocessed ByteBuffers received from the flow API.
+    private final BlockingQueue<List<ByteBuffer>> buffers;
+    private volatile Flow.Subscription subscription;
+    private volatile boolean subscribed;
+    private volatile boolean closed;
+    private volatile Throwable failed;
+    private volatile Iterator<ByteBuffer> currentListItr;
+    private volatile ByteBuffer currentBuffer;
+
+    private final Semaphore semaphore = new Semaphore(1);
+
+    private final long readTimeout;
+    private TimerTask currentTimeout;
+
+    /**
+     * Creates a response body subscriber that emits the response body as a {@link InputStream} while tracking a timeout
+     * for each value emitted by the subscription.
+     *
+     * @param readTimeout The timeout for reading each value emitted by the subscription.
+     */
+    public InputStreamTimeoutResponseSubscriber(long readTimeout) {
```

Comment:
nit: maybe int? I'm scared of timeouts that are longer than max int

Reply:
For simplicity, I'd rather leave it as `long` as that is what `Duration.toMillis()` returns and `int` doesn't really do much protection either as `Integer.MAX_VALUE` is still longer than anything reasonable.